File: solutions/alien-language/scala/AlienLang.scala
object AlienLanguage extends App {
  val source = scala.io.Source.fromFile(args(0))
  val lines = source.getLines.filter(_.length > 0)

  var params = lines.next.toString.split(" ").map(_.toInt)
  var dictionary_length = params(1).toInt

  //println(dictionary_length)
  var dictionary =   Array.empty[String]

  //Filling the dictrionary
  for (i <- 0 to dictionary_length - 1) {
    dictionary :+= lines.next.toString
  }

  var num_case = 0

  //Iterating over the test cases
  while(lines.hasNext){
    var test_case = lines.next //getting the regex
    test_case = test_case.toString.replace('(', '[')
    test_case = test_case.toString.replace(')', ']')

    var regex = test_case
    var res = 0

    //Matching de regex with each word of the dictionary
    for (i <- dictionary) {
      if (i.matches(regex)) { res += 1 }
    }

    num_case += 1   //Getting the testcase number
    println("Case #" + num_case  + ": " + res)
  }
}

File: solutions/alien-language/go/AlienLang.go
package main

import (
    "fmt"
    "log"
    "bufio"
    "strings"
    "strconv"
    "regexp"
    "os")

func main() {
    file, err := os.Open(os.Args[1])
    if err != nil {
        log.Fatal(err)
    }
    defer file.Close()

    scanner := bufio.NewScanner(file)
    var lines[]string
    for scanner.Scan() {
        lines = append(lines, scanner.Text())
    }

    params := strings.Split(lines[0], " ")
    dictionary_length, _ := strconv.Atoi(params[1])

    //Filling up the dictionary
    var dictionary[]string

    for i := 1;
    i < dictionary_length + 1;
    i += 1 {
        dictionary = append(dictionary, lines[i])
    }


    //Running the test cases
    for i := dictionary_length + 1;
    i < len(lines);
    i++{
        var test_case string = lines[i]
        var res int = 0
        test_case = strings.Replace(test_case, "(", "[", -1)
        test_case = strings.Replace(test_case, ")", "]", -1)


        regex, _ := regexp.Compile(test_case)

        for j:= 0;
        j < len(dictionary);
        j++{
            if regex.MatchString(dictionary[j]) == true {
                res += 1
            }



        }
        fmt.Println("Case #" + strconv.Itoa(i - len(dictionary)) + ": " + strconv.Itoa(res))
    }
}
